src/Simulator/IncrementalRedstoneSimulator/IncrementalRedstoneSimulator.cpp

```diff
@@ -0,0 +1,112 @@
+
+
+#include "Globals.h"
+
+#include "IncrementalRedstoneSimulator.h"
+#include "RedstoneHandlerFactory.h"
+#include "Chunk.h"
+
+
+
+
+
+void cIncrementalRedstoneSimulator::SimulateChunk(std::chrono::milliseconds a_Dt, int a_ChunkX, int a_ChunkZ, cChunk * a_Chunk)
+{
+	return;
+}
+
+
+
+
+
+void cIncrementalRedstoneSimulator::Simulate(float a_dt)
+{
+	for (auto & DelayInfo : m_Data.m_MechanismDelays)
+	{
+		std::get<1>(DelayInfo.second)++;
+
+		int DelayTicks, ElapsedTicks;
+		std::tie(DelayTicks, ElapsedTicks, std::ignore) = DelayInfo.second;
+		if (DelayTicks == ElapsedTicks)
+		{
+			m_Data.GetActiveBlocks().emplace_back(DelayInfo.first);
+		}
+	}
+
+	// Build our work queue
+	cVector3iArray WorkQueue;
+	std::swap(WorkQueue, m_Data.GetActiveBlocks());
+
+	// Process the work queue
+	while (!WorkQueue.empty())
+	{
+		// Grab the first element and remove it from the list
+		Vector3i CurrentLocation = WorkQueue.back();
+		WorkQueue.pop_back();
+
+		BLOCKTYPE CurrentBlock;
+		NIBBLETYPE CurrentMeta;
+		if (!m_World.GetBlockTypeMeta(CurrentLocation.x, CurrentLocation.y, CurrentLocation.z, CurrentBlock, CurrentMeta))
+		{
+			continue;
+		}
+
+		auto CurrentHandler = cRedstoneHandlerFactory::GetComponent(m_World, CurrentBlock, &m_Data).get();
+		if (CurrentHandler == nullptr)
+		{
+			continue;
+		}
+
+		cRedstoneHandler::PoweringData Power;
+		for (const auto & Location : CurrentHandler->GetValidSourcePositions(CurrentLocation, CurrentBlock, CurrentMeta))
```

Comment:
Is there a way to avoid the dynamic allocation in using a vector for GetValidSourcePostions's return. Perhaps a std::array of bools which can used against a static list?